File: 백준/Gold/1987. 알파벳/알파벳.cc
#include<bits/stdc++.h>
using namespace std;
int n, m, ret, visited[30]; char a[24][24];
const int dy[] = {-1, 0, 1, 0}, dx[] = {0, 1, 0, -1};
void go(int y, int x, int cnt){
    ret = max(ret, cnt);
    for(int i = 0; i < 4; i++){
        int ny = y + dy[i];
        int nx = x + dx[i];
        if(ny < 0 || nx < 0 || ny >= n || nx >= m) continue;
        int next = a[ny][nx] - 'A';
        if(visited[next] == 0){
            visited[next] = 1;
            go(ny, nx, cnt + 1);
            visited[next] = 0;
        }
    }
    return;
}
int main(){
    ios::sync_with_stdio(0);cin.tie(0);cout.tie(0);
    cin >> n >> m;
    for(int i = 0; i < n; i++){
        for(int j = 0; j < m; j++){
            cin >> a[i][j];
        }   
    }
    visited[a[0][0] - 'A'] = 1;
    go(0, 0, 1);
    cout << ret << '\n';
    return 0;
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.util.Scanner;

public class Main {
    static int n, m, ret;
    static int[] visited = new int[100];
    static char[][] a = new char[21][21];
    static final int[] dy = { -1, 0, 1, 0 };
    static final int[] dx = { 0, 1, 0, -1 };

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        n = scanner.nextInt();
        m = scanner.nextInt();

        for (int i = 0; i < n; i++) {
            String row = scanner.next();
            for (int j = 0; j < m; j++) {
                a[i][j] = row.charAt(j);
            }
        }

        visited[a[0][0]] = 1;
        go(0, 0, 1);

        System.out.println(ret);
    }

    static void go(int y, int x, int cnt) {
        ret = Math.max(ret, cnt);
        for (int i = 0; i < 4; i++) {
            int ny = y + dy[i];
            int nx = x + dx[i];
            if (ny < 0 || nx < 0 || ny >= n || nx >= m) {
                continue;
            }
            if (visited[a[ny][nx]] != 0) {
                continue;
            }
            visited[a[ny][nx]] = 1;
            go(ny, nx, cnt + 1);
            visited[a[ny][nx]] = 0;
        }
    }
}
